In [2]:
!pip install SQLAlchemy==1.4.4

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.5/7.5 MB 24.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for SQLAlchemy: filename=SQLAlchemy-1.4.4-cp310-cp310-linux_x86_64.whl size=1513275 sha256=4f3363c55ee707292c9b2e6293400cae54fa0c5c7164b7bfc1b0fc0cf9dcd5ec
  Stored in directory: /root/.cache/pip/wheels/af/c3/55/803261c43577195a922fd6aead99e6b4ed9be457dee35a1600
Successfully built SQLAlchemy
  Attempting uninstall: SQLAlchemy
    Found existing installation: SQLAlchemy 2.0.10
    Uninstalling SQLAlchemy-2.0.10:
      Successfully uninstalled SQLAlchemy-2.0.10


In [3]:
!pip install mysqlclient

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.1/88.1 kB 2.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for mysqlclient: filename=mysqlclient-2.1.1-cp310-cp310-linux_x86_64.whl size=108351 sha256=ec1bb9b69c7ae888f84d1c98fdd066901797ab731d85eebd0ff2ff486dad8ee6
  Stored in directory: /root/.cache/pip/wheels/7d/34/ba/a769c165b01646816afdf9bf792e847ef149693fee432b6b65
Successfully built mysqlclient


In [4]:
from sqlalchemy import create_engine

Load SQL extension 

In [5]:
%load_ext sql

Connect to the medical_devices database

In [6]:
%sql mysql://admin:isba_4715@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices

## Count of devices that do not have a version number:

In [7]:
%%sql
WITH cte_version_number AS (
	SELECT 
		model_number, 
		CASE WHEN version_number = 0 THEN 'N/A' ELSE version_number END AS version_number, 
		COUNT(*) AS VersionNumCount FROM device
	GROUP BY model_number, version_number
	ORDER BY version_number
)
SELECT COUNT(*) AS DevicesWithNoVersionNumber
FROM cte_version_number
WHERE version_number = 'N/A';

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
1 rows affected.


[(600,)]

This information can be useful for businesses that need to track or report on the status of their devices. A company that provides support for devices may need to know how many devices are running without a version number to ensure that they are up-to-date and secure. The absence of a version number may indicate that a device is an older model, and may have been discontinued or replaced by a newer version. Sales and customer support teams can use this information to provide accurate information to customers who may be interested in purchasing or servicing these devices.


# Number of Non-Sterile Devices by Brand with No Storage/Handling Information

In [35]:
%%sql
SELECT 
  company_name,
  brand_name, 
  COUNT(brand_name) AS brand_name_count
FROM device d
LEFT JOIN device_environment_conditions denvc 
    ON d.id = denvc.device_id
WHERE d.id NOT IN 
    (
        SELECT dt.device_id
        FROM device_terms dt
        WHERE dt.name NOT LIKE '%non-sterile%'
    )
    AND denvc.storage_handling_type IS NULL
GROUP BY brand_name
ORDER BY brand_name_count DESC;

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
34 rows affected.


[('GOLD STANDARD ORTHOPEDICS, LLC', 'GSO GS1 Cannulated Screw System', 230),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'GSO PCT Spinal System', 93),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'GSO ACP1 Anterior Cervical Plate System', 56),
 ('G SURGICAL LLC', 'GPS Anterior Cervical Plate System', 3),
 ('ZIRKONZAHN SRL', 'Dongle', 3),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KC 2124 Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 75-Y Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 105-2222 Swivel Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 105-2X22D Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 117-W Disposable Tee Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 120-AKW 45 degree Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 125-1TD Disposable Tee Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 169 Oxygen Tank Wrench with Security Cable', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'BE 169-W Oxygen Tank Wrench', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'DS 7-45  Bronchoscope Swivel', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KC 1010 Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KC 1515 Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KC 18901-K Elbow Connector ', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KC 2119 Reducing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 75-W Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KC 2126 Reducing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'KM 030 Disposable Connector with LCD Thermometer', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SG 101-D Reducing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 7-2 Five-in-One Connector', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 74-BK  Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 74-BL Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 74-C Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 74-G Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 74-W Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 74-Y Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 75-BK Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 75-BL Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 75-C Barbed Tubing Adapter', 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'SH 75-G Barbed Tubing Adapter', 1)]

A company could use this information to assess its own position within the market relative to its competitors. They could also use this information to identify potential areas for investment or development in terms of new devices that address gaps in the market or improve upon existing products. This can also help evaluate areas for improvement or alternative solutions.

##Devices that have missing sterilization information:

In [33]:
%%sql
SELECT id, brand_name, company_name, device_description
FROM device d
WHERE d.id NOT IN 
	(
	SELECT ds.device_id 
	FROM device_sterilization ds
	);

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
3 rows affected.


[(1, 'Dongle', 'ZIRKONZAHN SRL', 'Dongle for Scanner S600 ARTI\n'),
 (2, 'Dongle', 'ZIRKONZAHN SRL', 'Dongle for Scanner S600 ARTI\n'),
 (3, 'Dongle', 'ZIRKONZAHN SRL', 'Dongle for Scanner S600 ARTI\n')]

This information can be useful for businesses that need to ensure that their devices are properly sterilized to prevent the spread of infections or diseases. A hospital or medical facility can use this query to identify which devices need to be sterilized before being used on patients. Businesses can also ensure that they are in compliance with regulations and avoid potential legal or regulatory issues. By tracking device sterilization data, businesses can also ensure that their devices are properly maintained and serviced to help prevent device failures and improve overall device performance.

## Device Count and RX/OTC Breakdown by Product Code Name











In [39]:
%%sql
SELECT 
	dpc.product_code_name,
	SUM(d.is_rx) AS RX, 
	SUM(d.is_otc) AS OTC,
	COUNT(*) AS Total_product_code_name
FROM device_product_codes dpc 
JOIN device d 
	ON dpc.device_id = d.id
GROUP BY product_code_name 
ORDER BY COUNT(*) DESC;

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
112 rows affected.


[('Appliance, Fixation, Spinal Interlaminal', Decimal('213'), Decimal('0'), 214),
 ('Set, Tubing And Support, Ventilator (W Harness)', Decimal('0'), Decimal('159'), 159),
 ('Orthosis, Spinal Pedicle Fixation', Decimal('107'), Decimal('0'), 108),
 ('Syringe, Piston', Decimal('0'), Decimal('0'), 94),
 ('Needle, Hypodermic, Single Lumen', Decimal('0'), Decimal('0'), 80),
 ('Subsystem, Water Purification', Decimal('65'), Decimal('0'), 65),
 ('Appliance, Fixation, Spinal Intervertebral Body', Decimal('58'), Decimal('0'), 59),
 ('Syringe, Antistick', Decimal('0'), Decimal('0'), 43),
 ('Circuit, Breathing (W Connector, Adaptor, Y Piece)', Decimal('0'), Decimal('42'), 42),
 ('Orthopedic Manual Surgical Instrument', Decimal('42'), Decimal('0'), 42),
 ('Material, Polytetrafluoroethylene Vitreous Carbon, For Maxillofacial Reconstruction', Decimal('32'), Decimal('0'), 32),
 ('Tubing, Pressure And Accessories', Decimal('0'), Decimal('31'), 31),
 ('Filler, Bone Void, Calcium Compound', Decimal('28'), Decimal('0'), 28),
 ('Rod, Fixation, Intramedullary And Accessories, In-Vivo Cured, Light-Activated', Decimal('25'), Decimal('0'), 25),
 ('Valve, Non-Rebreathing', Decimal('16'), Decimal('0'), 16),
 ('Material, Tooth Shade, Resin', Decimal('0'), Decimal('0'), 15),
 ('Monitor, Airway Pressure (Includes Gauge And/Or Alarm)', Decimal('15'), Decimal('0'), 15),
 ('Abutment, Implant, Dental, Endosseous', Decimal('14'), Decimal('0'), 14),
 ('Material, Impression', Decimal('0'), Decimal('0'), 13),
 ('System, Test, Blood Glucose, Over The Counter', Decimal('0'), Decimal('6'), 13),
 ('Nebulizer (Direct Patient Interface)', Decimal('12'), Decimal('0'), 12),
 ('Connector, Airway (Extension)', Decimal('0'), Decimal('11'), 11),
 ('Meter, Airway Pressure (Inspiratory Force)', Decimal('11'), Decimal('0'), 11),
 ('Laryngoscope, Rigid', Decimal('0'), Decimal('11'), 11),
 ('Bag, Reservoir', Decimal('0'), Decimal('4'), 10),
 ('Changer, Tube, Endotracheal', Decimal('10'), Decimal('0'), 10),
 ('Crystal Violet For Histology', Decimal('0'), Decimal('0'), 9),
 ('Stimulator, Neuromuscular, External Functional', Decimal('9'), Decimal('0'), 9),
 ('Support, Breathing Tube', Decimal('0'), Decimal('9'), 9),
 ('Ventilator, Emergency, Manual (Resuscitator)', Decimal('9'), Decimal('0'), 9),
 ('Crown And Bridge, Temporary, Resin', Decimal('2'), Decimal('0'), 8),
 ('Congo Red', Decimal('0'), Decimal('0'), 8),
 ('Polymer Patient Examination Glove', Decimal('0'), Decimal('0'), 8),
 ('Airway, Oropharyngeal, Anesthesiology', Decimal('0'), Decimal('8'), 8),
 ('Methyl Methacrylate For Cranioplasty', Decimal('7'), Decimal('0'), 7),
 ('Sterilization Wrap Containers, Trays, Cassettes & Other Accessories', Decimal('0'), Decimal('0'), 7),
 ('Tray, Surgical', Decimal('7'), Decimal('0'), 7),
 ('Filler, Bone Void, Osteoinduction (W/O Human Growth Factor)', Decimal('6'), Decimal('0'), 6),
 ('Photocoagulator And Accessories', Decimal('6'), Decimal('0'), 6),
 ('Tube, Tracheal (W/Wo Connector)', Decimal('0'), Decimal('0'), 6),
 ('Tap, Bone', Decimal('6'), Decimal('0'), 6),
 ('Accelerator, Linear, Medical', Decimal('5'), Decimal('0'), 5),
 ('Mesh, Surgical, Polymeric', Decimal('5'), Decimal('0'), 5),
 ('Instrument, Bending Or Contouring', Decimal('5'), Decimal('0'), 5),
 ('Microscope, Surgical', Decimal('0'), Decimal('0'), 4),
 ('Accessories, Cleaning, For Endoscope', Decimal('0'), Decimal('4'), 4),
 ('Biebrich Scarlet', Decimal('0'), Decimal('0'), 4),
 ('System, Measurement, Blood-Pressure, Non-Invasive', Decimal('0'), Decimal('0'), 4),
 ('Ventilator, Emergency, Powered (Resuscitator)', Decimal('4'), Decimal('0'), 4),
 ('Implant, Endosseous, Root-Form', Decimal('2'), Decimal('0'), 3),
 ('System, Environmental Control, Powered', Decimal('0'), Decimal('0'), 3),
 ('Bone Grafting Material, Human Source', Decimal('3'), Decimal('0'), 3),
 ('Denture, Plastic, Teeth', Decimal('3'), Decimal('0'), 3),
 ('Dental Barriers And Sleeves', Decimal('0'), Decimal('0'), 3),
 ('Stain, Papanicolau', Decimal('0'), Deci

Each product category either is available only as presciption or only over-the-counter. Product management teams can use this to identify which product code names (categories) have the most prescription and OTC devices. This  can help companies in making strategic decisions on the products to produce and the distribution channels to use. It can also help companies find any patterns in their product offerings that may be more profitable or have higher demand.

### Count of Devices with No Version Number

This information can be useful for product management teams who are looking to update and maintain their product catalog. 

In [13]:
%%sql 
WITH cte_version_number AS (
	SELECT model_number, CASE WHEN version_number = 0 THEN 'N/A' ELSE version_number END AS version_number, COUNT(*) AS VersionNumCount FROM device
	GROUP BY model_number, version_number
	ORDER BY version_number
)
SELECT COUNT(*) AS DevicesWithNoVersionNumber
FROM cte_version_number
WHERE version_number = 'N/A';

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
1 rows affected.


[(600,)]

# Business Problem:

## A company is interested in developing a new device to help spinal cord patients recover more comfortably. They want to evaluate the competitiveness of their proposed product against available products on the market. They want to know which devices have special handling conditions, such as requiring a sterile environment or specific storage conditions. They also want to know the frequency of updates made to products in this category.


## Average Number of Updates By Company Per Year


In [40]:
%%sql
-- looking at average updates per year by company
-- for their device offerings, to evaluate the need for a new product on the market
WITH cte_publish_data AS (
	SELECT 
		company_name, 
		year(publish_date) AS publish_year, 
		month(publish_date) AS publish_month, 
		COUNT(*) as publish_count
	FROM device
	GROUP BY company_name, year(publish_date), month(publish_date)
), 
cte_publish_years AS (
	SELECT 
		company_name,
		publish_year,
		SUM(publish_count) AS updates_year
	FROM cte_publish_data
	GROUP BY company_name, publish_year
)
SELECT 
	company_name, 
	ROUND(SUM(updates_year) / COUNT(*), 0) AS avg_updates_per_year
FROM cte_publish_years 
GROUP BY company_name
ORDER BY avg_updates_per_year DESC;

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
46 rows affected.


[('GOLD STANDARD ORTHOPEDICS, LLC', Decimal('113')),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', Decimal('53')),
 ('MHC MEDICAL PRODUCTS, LLC', Decimal('47')),
 ('TRANSENE COMPANY, INC.', Decimal('46')),
 ('BETTER WATER, LLC', Decimal('33')),
 ('CEREMED, INC.', Decimal('32')),
 ('Musculoskeletal Transplant Foundation, Inc.', Decimal('14')),
 ('ILLUMINOSS MEDICAL, INC.', Decimal('13')),
 ('TOP QUALITY MANUFACTURING, INC.', Decimal('11')),
 ('NATHARN TRADING COMPANY LIMITED', Decimal('8')),
 ('VORTEX SURGICAL, LLC', Decimal('8')),
 ('WAL-MART STORES, INC.', Decimal('8')),
 ('Hager & Meisinger Gesellschaft mit beschränkter Haftung', Decimal('7')),
 ('BRYAN MEDICAL, INC.', Decimal('6')),
 ('RESTORATIVE THERAPIES, INC.', Decimal('5')),
 ('MEDLINX ACACIA PTE. LTD.', Decimal('5')),
 ('Ecleris', Decimal('4')),
 ('CHAMELEON DENTAL PRODUCTS, INC', Decimal('4')),
 ('ZIRKONZAHN SRL', Decimal('4')),
 ('SHIFENG MEDICAL TECHNOLOGY CO.，LTD.', Decimal('3')),
 ('Crest Healthcare Supply', Decimal('3')),
 ('RADIOMED CORP', Decimal('3')),
 ('Vivorte, Inc', Decimal('3')),
 ('G SURGICAL LLC', Decimal('3')),
 ('bredent GmbH & Co. KG', Decimal('2')),
 ('CUSTOM ULTRASONICS, INCORPORATED', Decimal('2')),
 ('INTRA-LOCK INTERNATIONAL, INC.', Decimal('2')),
 ('OPA DENTAL LLC', Decimal('2')),
 ('Braebon Medical Corporation', Decimal('1')),
 ('CMP INDUSTRIES LLC', Decimal('1')),
 ('VUCOMP, INC.', Decimal('1')),
 ('TELE-MADE DISPOSABLES, INC.', Decimal('1')),
 ('PARK DENTAL RESEARCH CORP', Decimal('1')),
 ('NAVILYST MEDICAL, INC.', Decimal('1')),
 ('PANTHEON SPINAL, LLC', Decimal('1')),
 ('NORTH EAST SCIENTIFIC INC', Decimal('1')),
 ('POLYTECHNIC RESOURCES INC', Decimal('1')),
 ('LANG DENTAL MANUFACTURING CO., INC.', Decimal('1')),
 ('COSMEDENT, INC.', Decimal('1')),
 ('Shukla Medical', Decimal('1')),
 ('MEDIVIS, INC', Decimal('1')),
 ('M.A.C. MEDICAL INC.', Decimal('1')),
 ('STRADIS MEDICAL, LLC', Decimal('1')),
 ('EYEKON MEDICAL, INC.', Decimal('1')),
 ('VERICEL CORPORATION', Decimal('1')),
 ('Cora', Decimal('1'))]

This shows that Gold Standard Orthopedics is the biggest competitor for this company that is trying to create a new product. This company has all of its brands with high quality products having many updates per year to keep its products up to date. Companies that release a higher number of updates per year may be more dedicated to improving their products and keeping up with the latest technologies, which can be an advantage in a competitive market.
The average number of updates per year for each company can help product managers and investors to evaluate the company's product development cycle and the rate of innovation compared to the competition.
If a company's average number of updates per year is significantly lower than its competitors, it may indicate that the company is struggling to keep up with the market demands and needs to invest more in its product development and innovation.

## Frequency of updates made to products in each category


In [48]:
%%sql 
SELECT 	
	company_name, 
  dpc.product_code_name, 
	year(publish_date) AS Year, 
  MONTH(publish_date) AS Month,
	COUNT(*) OVER (PARTITION BY YEAR(publish_date), MONTH(publish_date)) AS update_frequency_count
FROM device d 
JOIN device_product_codes dpc 
	ON dpc.device_id = d.id 
GROUP BY dpc.product_code_name, YEAR(publish_date), MONTH(publish_date);

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
183 rows affected.


[('VUCOMP, INC.', 'Analyzer,Medical Image', 2015, 2, 1),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Cuff, Tracheal Tube, Inflatable', 2015, 7, 1),
 ('Hager & Meisinger Gesellschaft mit beschränkter Haftung', 'Implant, Endosseous, Root-Form', 2015, 8, 2),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'Appliance, Fixation, Spinal Interlaminal', 2015, 8, 2),
 ('EYEKON MEDICAL, INC.', 'Intraocular Lens', 2015, 9, 2),
 ('PARK DENTAL RESEARCH CORP', 'Implant, Endosseous, Root-Form', 2015, 9, 2),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Bag, Reservoir', 2015, 10, 6),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Ventilator, Continuous, Facility Use', 2015, 10, 6),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Valve, Non-Rebreathing', 2015, 10, 6),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Ventilator, Emergency, Manual (Resuscitator)', 2015, 10, 6),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Ventilator, Emergency, Powered (Resuscitator)', 2015, 10, 6),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Attachment, Breathing, Positive End Expiratory Pressure', 2015, 10, 6),
 ('MEDLINX ACACIA PTE. LTD.', 'Mesh, Surgical, Polymeric', 2015, 12, 2),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'Appliance, Fixation, Spinal Interlaminal', 2015, 12, 2),
 ('Vivorte, Inc', 'Filler, Bone Void, Osteoinduction (W/O Human Growth Factor)', 2016, 1, 1),
 ('Musculoskeletal Transplant Foundation, Inc.', 'Methyl Methacrylate For Cranioplasty', 2016, 2, 6),
 ('Musculoskeletal Transplant Foundation, Inc.', 'Filler, Bone Void, Calcium Compound', 2016, 2, 6),
 ('Musculoskeletal Transplant Foundation, Inc.', 'Filler, Bone Void, Osteoinduction (W/O Human Growth Factor)', 2016, 2, 6),
 ('Musculoskeletal Transplant Foundation, Inc.', 'Bone Grafting Material, Human Source', 2016, 2, 6),
 ('STRADIS MEDICAL, LLC', 'Unit, Operative Dental', 2016, 2, 6),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Nebulizer (Direct Patient Interface)', 2016, 2, 6),
 ('STRADIS MEDICAL, LLC', 'Kit, Plastic Surgery And Accessories', 2016, 5, 3),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Monitor, Airway Pressure (Includes Gauge And/Or Alarm)', 2016, 5, 3),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'Guide, Surgical, Instrument', 2016, 5, 3),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Nebulizer (Direct Patient Interface)', 2016, 6, 2),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Monitor, Airway Pressure (Includes Gauge And/Or Alarm)', 2016, 6, 2),
 ('bredent GmbH & Co. KG', 'Denture, Plastic, Teeth', 2016, 7, 5),
 ('CEREMED, INC.', 'Material, Polytetrafluoroethylene Vitreous Carbon, For Maxillofacial Reconstruction', 2016, 7, 5),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Meter, Airway Pressure (Inspiratory Force)', 2016, 7, 5),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'Orthosis, Spinal Pedicle Fixation', 2016, 7, 5),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'Appliance, Fixation, Spinal Interlaminal', 2016, 7, 5),
 ('MHC MEDICAL PRODUCTS, LLC', 'Needle, Hypodermic, Single Lumen', 2016, 8, 8),
 ('MHC MEDICAL PRODUCTS, LLC', 'Syringe, Antistick', 2016, 8, 8),
 ('MHC MEDICAL PRODUCTS, LLC', 'Syringe, Piston', 2016, 8, 8),
 ('MHC MEDICAL PRODUCTS, LLC', 'System, Measurement, Blood-Pressure, Non-Invasive', 2016, 8, 8),
 ('MHC MEDICAL PRODUCTS, LLC', 'System, Test, Blood Glucose, Over The Counter', 2016, 8, 8),
 ('MHC MEDICAL PRODUCTS, LLC', 'Single (Specified) Analyte Controls (Assayed And Unassayed)', 2016, 8, 8),
 ('INSTRUMENTATION INDUSTRIES, INCORPORATED', 'Nebulizer (Direct Patient Interface)', 2016, 8, 8),
 ('GOLD STANDARD ORTHOPEDICS, LLC', 'Appliance, Fixation, Spinal Intervertebral Body', 2016, 8, 8),
 ('Hager & Meisinger Gesellschaft mit beschränkter Haftung', 'Abutment, Implant, Dental, Endosseous', 2016, 9, 13),
 ('CUSTOM ULTRASONICS, INCORPORATED', 'Accessories, Cleaning, For Endoscope', 2016, 9, 13),
 ('RESTORATIVE THERAPIES, INC.', 'Stimulator, Neuromuscular, External Functional', 2016, 9, 13),
 ('Braebon Medical Corporation', 'Ventilatory Effort Recorder', 2016, 9, 13),
 ('M

## Saturation of products with prescription vs OTC

In [53]:
%%sql
SELECT 
 dpc.product_code_name, 
    CASE WHEN d.is_rx = 1 THEN 'Prescription Required' ELSE 'No prescription required' END AS is_prescription_required, 
    CASE WHEN d.is_otc = 1 THEN 'Available over the counter' ELSE 'Not available over the counter' END AS is_over_the_counter, 
    COUNT(*) 
FROM device d
JOIN device_product_codes dpc on d.id = dpc.device_id
GROUP BY dpc.product_code_name, d.is_rx, d.is_otc
ORDER BY COUNT(*) DESC;

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
122 rows affected.


[('Appliance, Fixation, Spinal Interlaminal', 'Prescription Required', 'Not available over the counter', 213),
 ('Set, Tubing And Support, Ventilator (W Harness)', 'No prescription required', 'Available over the counter', 159),
 ('Orthosis, Spinal Pedicle Fixation', 'Prescription Required', 'Not available over the counter', 107),
 ('Syringe, Piston', 'No prescription required', 'Not available over the counter', 94),
 ('Needle, Hypodermic, Single Lumen', 'No prescription required', 'Not available over the counter', 80),
 ('Subsystem, Water Purification', 'Prescription Required', 'Not available over the counter', 65),
 ('Appliance, Fixation, Spinal Intervertebral Body', 'Prescription Required', 'Not available over the counter', 58),
 ('Syringe, Antistick', 'No prescription required', 'Not available over the counter', 43),
 ('Circuit, Breathing (W Connector, Adaptor, Y Piece)', 'No prescription required', 'Available over the counter', 42),
 ('Orthopedic Manual Surgical Instrument', 'Prescription Required', 'Not available over the counter', 42),
 ('Material, Polytetrafluoroethylene Vitreous Carbon, For Maxillofacial Reconstruction', 'Prescription Required', 'Not available over the counter', 32),
 ('Tubing, Pressure And Accessories', 'No prescription required', 'Available over the counter', 31),
 ('Filler, Bone Void, Calcium Compound', 'Prescription Required', 'Not available over the counter', 28),
 ('Rod, Fixation, Intramedullary And Accessories, In-Vivo Cured, Light-Activated', 'Prescription Required', 'Not available over the counter', 25),
 ('Valve, Non-Rebreathing', 'Prescription Required', 'Not available over the counter', 16),
 ('Material, Tooth Shade, Resin', 'No prescription required', 'Not available over the counter', 15),
 ('Monitor, Airway Pressure (Includes Gauge And/Or Alarm)', 'Prescription Required', 'Not available over the counter', 15),
 ('Abutment, Implant, Dental, Endosseous', 'Prescription Required', 'Not available over the counter', 14),
 ('Material, Impression', 'No prescription required', 'Not available over the counter', 13),
 ('Nebulizer (Direct Patient Interface)', 'Prescription Required', 'Not available over the counter', 12),
 ('Connector, Airway (Extension)', 'No prescription required', 'Available over the counter', 11),
 ('Meter, Airway Pressure (Inspiratory Force)', 'Prescription Required', 'Not available over the counter', 11),
 ('Laryngoscope, Rigid', 'No prescription required', 'Available over the counter', 11),
 ('Changer, Tube, Endotracheal', 'Prescription Required', 'Not available over the counter', 10),
 ('Crystal Violet For Histology', 'No prescription required', 'Not available over the counter', 9),
 ('Stimulator, Neuromuscular, External Functional', 'Prescription Required', 'Not available over the counter', 9),
 ('Support, Breathing Tube', 'No prescription required', 'Available over the counter', 9),
 ('Ventilator, Emergency, Manual (Resuscitator)', 'Prescription Required', 'Not available over the counter', 9),
 ('Congo Red', 'No prescription required', 'Not available over the counter', 8),
 ('Polymer Patient Examination Glove', 'No prescription required', 'Not available over the counter', 8),
 ('Airway, Oropharyngeal, Anesthesiology', 'No prescription required', 'Available over the counter', 8),
 ('Methyl Methacrylate For Cranioplasty', 'Prescription Required', 'Not available over the counter', 7),
 ('Sterilization Wrap Containers, Trays, Cassettes & Other Accessories', 'No prescription required', 'Not available over the counter', 7),
 ('System, Test, Blood Glucose, Over The Counter', 'No prescription required', 'Not available over the counter', 7),
 ('Tray, Surgical', 'Prescription Required', 'Not available over the counter', 7),
 ('Filler, Bone Void, Osteoinduction (W/O Human Growth Factor)', 'Prescription Required', 'Not available over the counter', 6),
 ('Photocoagulator And Accessories', 'Prescription Required', 'Not available over the counter', 6),
 ('Crown And Bridge, Temporary, Resi

The resulting table shows Spinal cord devices in the top 3. With spinal inter laminal appliance at the highest with 213 products, and spinal pedicle fixation orthosis with 107 products. They both obviously are prescription only products. This shows that understand the spinal devices are in a highly competitive landscape of available products, especially with the the level of regulation(where earlier we saw that they highly produced as sterile products with proper instructions, and high update frequency. This indicates that it has a high market demand. 

## View of Full Device Data

In [61]:
%%sql
CREATE VIEW vw_full_device_data
AS
SELECT
 d.brand_name, d.version_status, d.record_status, d.version_number, d.version_date, d.publish_date, d.distribution_end_date, d.distribution_status, d.model_number,
    d.company_name, d.device_description, d.is_rx, d.is_otc,
    dc.email_address, dc.phone_number,
    denvc.storage_handling_type, storage_handling_special_condition_text,
    di.device_identifier, di.issuing_agency, dms. submission_number, dms.supplement_number,
    dpc.product_code, dpc.product_code_name,
    ds.size_type, ds.size, ds.size_unit, ds.size_text,
    dst.is_device_sterile,
    dst.require_initixal_sterilization,
    dst.method,
    dt.name, dt.pt_definition
FROM device d
LEFT JOIN device_contacts dc 
	ON dc.device_id = d.id
LEFT JOIN device_environment_conditions denvc 
	ON denvc.device_id = d.id
LEFT JOIN device_identifiers di 
	ON di.device_id = d.id
LEFT JOIN device_market_submissions dms 
	ON dms.device_id = d.id
LEFT JOIN device_product_codes dpc 
	ON dpc.device_id = d.id
LEFT JOIN device_sizes ds on ds.device_id = d.id
LEFT JOIN device_sterilization dst 	
	ON dst.device_id = d.id
LEFT JOIN device_terms dt
	ON dt.device_id = d.id;

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
(MySQLdb.OperationalError) (1054, "Unknown column 'dst.require_initixal_sterilization' in 'field list'")
[SQL: CREATE VIEW vw_full_device_data
AS
SELECT
 d.brand_name, d.version_status, d.record_status, d.version_number, d.version_date, d.publish_date, d.distribution_end_date, d.distribution_status, d.model_number,
    d.company_name, d.device_description, d.is_rx, d.is_otc,
    dc.email_address, dc.phone_number,
    denvc.storage_handling_type, storage_handling_special_condition_text,
    di.device_identifier, di.issuing_agency, dms. submission_number, dms.supplement_number,
    dpc.product_code, dpc.product_code_name,
    ds.size_type, ds.size, ds.size_unit, ds.size_text,
    dst.is_device_sterile,
    dst.require_initixal_sterilization,
    dst.method,
    dt.name, dt.pt_definition
FROM device d
LEFT JOIN device_contacts dc 
	ON dc.device_id = d.id
LEFT JOIN device_environment_conditions denvc 

In [60]:
%%sql
SELECT *
FROM vw_full_device_data
LIMIT 10;

 * mysql://admin:***@isba-dev-01.cpkvznb9aubv.us-east-1.rds.amazonaws.com/medical_devices
10 rows affected.


[('Dongle', 'Update', 'Published', 0, datetime.datetime(2022, 2, 21, 0, 0), datetime.datetime(2018, 7, 6, 0, 0), datetime.datetime(1, 1, 1, 0, 0), 'In Commercial Distribution', 'ZBAC1401 ', 'ZIRKONZAHN SRL', 'Dongle for Scanner S600 ARTI\n', b'\x00', b'\x00', None, None, None, None, 'ZBAC1401', 'HIBCC', None, None, 'NOF', 'System, Optical Impression, Computer Assisted Design And Manufacturing (Cad/Cam) Of Dental Restorations', None, None, None, None, b'\x00', b'\x00', None, 'Dental CAD/CAM system, laboratory', 'An assembly of devices designed for automated computer-aided design (CAD) and computer-aided manufacturing (CAM) of custom-made (prescribed for one p ... (384 characters truncated) ... ing, simulation and device production; and robotic capabilities for production of the actual device(s) for delivery and final fitting to the patient.'),
 ('OCTAGON Infinity TL RP ALLIGATOR Abutment angled, 15°, H 3.0 mm', 'Update', 'Published', 5098008, datetime.datetime(2019, 11, 8, 0, 0), datetime.datetime(2016, 9, 8, 0, 0), datetime.datetime(1, 1, 1, 0, 0), 'Not in Commercial Distribution', '5098008', 'Hager & Meisinger Gesellschaft mit beschränkter Haftung', 'TL RP ALLIGATOR Abutment angled, 15°, H 3.0 mm for OCTAGON Infinity Dental Implant TL', b'\x01', b'\x00', 'kathystrange@acesurgical.com', '+1(508)588-3100', None, None, 'XXXX50980089', 'HIBCC', None, None, 'NHA', 'Abutment, Implant, Dental, Endosseous', None, None, None, None, b'\x00', b'\x01', 'Moist Heat or Steam Sterilization', 'Dental implant suprastructure, permanent, preformed', 'A prefabricated device intended to provide a permanent intermediate fixture level between a dental implant and the final prosthesis/restoration (e.g. ... (353 characters truncated) ... cluded, however the suprastructure does not represent the screw(s) in isolation. It may also be referred to as an abutment assembly or mesostructure.'),
 ('OCTAGON Infinity TL RP ALLIGATOR Abutment GH 1.0 mm', 'Update', 'Published', 5098001, datetime.datetime(2019, 11, 8, 0, 0), datetime.datetime(2016, 9, 8, 0, 0), datetime.datetime(1, 1, 1, 0, 0), 'Not in Commercial Distribution', '5098001', 'Hager & Meisinger Gesellschaft mit beschränkter Haftung', 'TL RP ALLIGATOR Abutment GH 1.0 mm for OCTAGON Infinity Dental Implant TL', b'\x01', b'\x00', 'kathystrange@acesurgical.com', '+1(508)588-3100', None, None, 'XXXX50980019', 'HIBCC', None, None, 'NHA', 'Abutment, Implant, Dental, Endosseous', None, None, None, None, b'\x00', b'\x01', 'Moist Heat or Steam Sterilization', 'Dental implant suprastructure, permanent, preformed', 'A prefabricated device intended to provide a permanent intermediate fixture level between a dental implant and the final prosthesis/restoration (e.g. ... (353 characters truncated) ... cluded, however the suprastructure does not represent the screw(s) in isolation. It may also be referred to as an abutment assembly or mesostructure.'),
 ('OCTAGON Infinity TL RP Abutment angled 15° B H= 5.7 mm cementable ', 'Update', 'Published', 5082649, datetime.datetime(2019, 11, 8, 0, 0), datetime.datetime(2016, 9, 8, 0, 0), datetime.datetime(1, 1, 1, 0, 0), 'Not in Commercial Distribution', '5082649', 'Hager & Meisinger Gesellschaft mit beschränkter Haftung', 'TL RP Abutment angled 15° B H= 5.7 mm cementable for OCTAGON Infinity Dental Implant RP', b'\x01', b'\x00', 'kathystrange@acesurgical.com', '+1(508)588-3100', None, None, 'XXXX50826499', 'HIBCC', None, None, 'NHA', 'Abutment, Implant, Dental, Endosseous', None, None, None, None, b'\x00', b'\x01', 'Moist Heat or Steam Sterilization', 'Dental implant suprastructure, permanent, preformed', 'A prefabricated device intended to provide a permanent intermediate fixture level between a dental implant and the final prosthesis/restoration (e.g. ... (353 characters truncated) ... cluded, however the suprastructure does not represent the screw(s) in isolation. It may also be referred to as an abutment assembly or mesostructure.'),
 ('OCTAGON Infinity TL RP Abutm

This view can help the company gain a comprehensive understanding of the medical devices on the market and monitor their performance across various dimensions such as size, sterilization, product codes, market submissions, etc. This view can also be used to track changes in the version number, status and date of a device, which can help them ensure regulatory compliance and manage recalls if necessary. Sales teams can use this view to identify devices that are available for sale, their distribution status and the contact information for the manufacturer, which can help them manage relationships with clients and customers.


# Recommendation

The company should focus on developing a high-quality, innovative product that addresses gaps in the market. They can make a product that works for spinal cord patients but they must be aware of high level competition, if they are not prepared for that then they can choose to make a medical device in a different category. They should also ensure that their device is sterile and has proper storage and handling instructions to compete in a market where these factors are important and they’d be competing with 93 other devices. But specifically 2 spinal cord devices that are very successful. Using the view, the company can inspect deeper into the data and see what other companies and brands that may be easier competition, or invest in developing an already existing device that is in low demand but help improve it to be up to standards to the market. Especially seeing the different companies succeed in 2019 and 2022, the company can use this view to see other information on those companies during those years to see what they did to have the success in those specific years and apply that knowledge when creating the product. In any device that the company decides to make, they should be dedicated to improving their products and keeping up with the latest technologies, to have the competitive advantage. 


# Conclusion

This project is a great sample of what I’d like to do at Kroll Government Solutions. This position requires working with various data sets and programming data import/export, performing analysis, which I did in this project as I collected data on medical devices by finding the API documentation on the national library of medicine website, and filtered the data to show which devices were sterile and had proper storage and handling instructions. This information can help product managers and investors in the company to evaluate the company's product development cycle and the rate of innovation compared to the competition, to help create the right product for success in the market.

